packages/vite/src/node/plugins/resolve.ts

```diff
@@ -649,10 +649,25 @@ export function tryNodeResolve(
   }
 
   let pkg: PackageData | undefined
-  const pkgId = possiblePkgIds.reverse().find((pkgId) => {
-    pkg = resolvePackageData(pkgId, basedir, preserveSymlinks, packageCache)!
-    return pkg
-  })!
+  let pkgId: string
+
+  const rootPkg =
+    possiblePkgIds.length &&
+    resolvePackageData(
+      possiblePkgIds[0],
+      basedir,
+      preserveSymlinks,
+      packageCache
+    )!
+  if (rootPkg && rootPkg?.data?.exports) {
+    pkg = rootPkg
+    pkgId = possiblePkgIds[0]
+  } else {
+    pkgId = possiblePkgIds.reverse().find((pkgId) => {
+      pkg = resolvePackageData(pkgId, basedir, preserveSymlinks, packageCache)!
+      return pkg
+    })!
+  }
```

Comment:
I wonder if this is correct to spec according to https://nodejs.org/api/esm.html#resolver-algorithm-specification.

<img width="914" alt="image" src="https://user-images.githubusercontent.com/34116392/195361980-1d3eec37-4fed-443e-9817-b72224001aff.png">

I think we would need to update `resolvePackageData` to find a `package.json` that first have `exports`. But also keep another variable for the closest `package.json` regardless to verify `type: module` IIUC 😵

Reply:
I don't think it'd make it to step 11. It looks like it'd stop at steps 9/10. That being said, you're right that this isn't quite correct. I'll update it

Reply:
fixed

Reply:
> What I found is that it will only use `exports` from the `package.json` in the package root,

Ah I thought the spec mentioned that it will take the closest `package.json` that has the first `exports`, but if you've tested that it will always take the root, then that sounds good too.